tools/cloudflow-operator/src/main/scala/cloudflow/operator/action/EndpointActions.scala

```diff
@@ -0,0 +1,152 @@
+    val newEndpoints: Set[App.Endpoint] = distinctEndpoints(newApp.spec)
+
+    val deleteActions = (currentEndpoints -- newEndpoints).flatMap { endpoint: App.Endpoint =>
+      Seq(
+        Action.delete[Service](
+          Name.ofService(StreamletDeployment.name(newApp.spec.appId, endpoint.streamlet.getOrElse("no-name"))),
+          newApp.namespace))
+    }.toList
+    val createActions = (newEndpoints -- currentEndpoints).flatMap { endpoint =>
+      Seq(
+        createServiceAction(
+          endpoint,
+          newApp,
+          StreamletDeployment.name(newApp.spec.appId, endpoint.streamlet.getOrElse("no-name"))))
+    }.toList
+    deleteActions ++ createActions
+  }
+
+  private def serviceResource(
+      endpoint: App.Endpoint,
+      streamletDeploymentName: String,
+      namespace: String,
+      labels: CloudflowLabels,
+      ownerReferences: List[OwnerReference]): Service = {
+    val servicePort = {
+      new ServicePortBuilder()
+        .withName(Name.ofContainerPort(endpoint.containerPort.getOrElse(DefaultContainerPort)))
+        .withPort(Integer.valueOf(endpoint.containerPort.getOrElse(DefaultContainerPort)))
+        .withTargetPort(
+          new IntOrStringBuilder()
+            .withNewStrVal(Name.ofContainerPort(endpoint.containerPort.getOrElse(DefaultContainerPort)))
+            .build())
+        .build()
+    }
+
+    new ServiceBuilder()
+      .withNewMetadata()
+      .withName(Name.ofService(streamletDeploymentName))
+      .withNamespace(namespace)
+      .withLabels(labels(Name.ofService(streamletDeploymentName)).asJava)
+      .withOwnerReferences(ownerReferences: _*)
+      .endMetadata()
+      .withSpec(
+        new ServiceSpecBuilder()
+          .withType("ClusterIP")
+          .withPorts(servicePort)
+          .withSelector(Map(CloudflowLabels.Name -> Name.ofPod(streamletDeploymentName)).asJava)
+          .build())
+      .build()
+  }
+
+  private def createServiceAction(endpoint: App.Endpoint, app: App.Cr, streamletDeploymentName: String): Action = {
+    val labels = CloudflowLabels(app)
+    val ownerReferences = List(AppOwnerReference(app.getMetadata().getName(), app.getMetadata().getUid()))
+
+    CreateServiceAction(serviceResource(endpoint, streamletDeploymentName, app.namespace, labels, ownerReferences))
+  }
+
+  /**
+   * Creates an action for creating a service.
+   */
+  case class CreateServiceAction(service: Service)(implicit val lineNumber: sourcecode.Line, val file: sourcecode.File)
+      extends Action {
+
+    val getOperation = { client: KubernetesClient =>
+      (client.services().asInstanceOf[MixedOperation[Service, ServiceList, Resource[Service]]])
+    }
+
+    val executeOperation = { services: MixedOperation[Service, ServiceList, Resource[Service]] =>
+      val thisService =
+        services.inNamespace(service.getMetadata.getNamespace).withName(service.getMetadata.getName)
+      val ser = thisService.get()
+      Option(ser) match {
```

Comment:
Maybe do `Option(thisService.get())` so no one can make the mistake to use ser (super minor)